src/vandermonde.jl

```diff
@@ -1,63 +1,85 @@
 export Vandermonde
+
 """
-[`Vandermonde` matrix](http://en.wikipedia.org/wiki/Vandermonde_matrix)
+    V = Vandermonde(c::AbstractVector)
 
-```julia
-julia> a = collect(1.0:5.0)
-julia> A = Vandermonde(a)
-5×5 Vandermonde{Float64}:
- 1.0  1.0   1.0    1.0    1.0
- 1.0  2.0   4.0    8.0   16.0
- 1.0  3.0   9.0   27.0   81.0
- 1.0  4.0  16.0   64.0  256.0
- 1.0  5.0  25.0  125.0  625.0
+Create a "lazy" `n × n`
+[`Vandermonde` matrix](http://en.wikipedia.org/wiki/Vandermonde_matrix)
+but requiring only `O(n)` storage for the vector `c`.
+
+```jldoctest van
+julia> a = 1:5; A = Vandermonde(a)
+5×5 Vandermonde{Int64, UnitRange{Int64}}:
+ 1  1   1    1    1
+ 1  2   4    8   16
+ 1  3   9   27   81
+ 1  4  16   64  256
+ 1  5  25  125  625
 ```
 
 Adjoint Vandermonde:
-```julia
+```jldoctest van
 julia> A'
-5×5 LinearAlgebra.Adjoint{Float64,Vandermonde{Float64}}:
- 1.0   1.0   1.0    1.0    1.0
- 1.0   2.0   3.0    4.0    5.0
- 1.0   4.0   9.0   16.0   25.0
- 1.0   8.0  27.0   64.0  125.0
- 1.0  16.0  81.0  256.0  625.0
+5×5 adjoint(::Vandermonde{Int64, UnitRange{Int64}}) with eltype Int64:
+ 1   1   1    1    1
+ 1   2   3    4    5
+ 1   4   9   16   25
+ 1   8  27   64  125
+ 1  16  81  256  625
 ```
 
-The backslash operator \\ is overloaded to solve Vandermonde and adjoint
+The backslash operator `\\` is overloaded to solve Vandermonde and adjoint
 Vandermonde systems in `O(n^2)` time using the algorithm of
 Björck & Pereyra (1970), https://doi.org/10.2307/2004623.
 
-```julia
-julia> A\a
-5-element Array{Float64,1}:
+```jldoctest van
+julia> A \\ a
+5-element Vector{Float64}:
  0.0
  1.0
  0.0
  0.0
  0.0
+```
 
-julia> r2 = A[2,:]
-julia> A'\r2
-5-element Array{Float64,1}:
+```jldoctest van
+julia> A' \\ A[2,:]
+5-element Vector{Float64}:
  0.0
  1.0
  0.0
  0.0
  0.0
 ```
 """
-struct Vandermonde{T} <: AbstractArray{T, 2}
-    c :: Vector{T}
+struct Vandermonde{T,C} <: AbstractMatrix{T}
+    c :: C
+
+    function Vandermonde(c::AbstractVector{T}) where T
+        axes(c,1) isa Base.OneTo || throw(ArgumentError("must be OneTo"))
```

Comment:
It would be nice to remove this restriction at some point.

Reply:
Thanks for all these helpful suggestions.  I've accepted all of them.
But I am unsure what one would do with an OffsetVector input here.

I got into the habit of checking axes after working with @dlfivefifty on some other packages.  So just now I checked to see what `ToeplitzMatrices` does with `AbstractVector` inputs:

https://github.com/JuliaMatrices/ToeplitzMatrices.jl/blob/edbc8eb37c93df6c752e5313da307445a2e2022c/src/ToeplitzMatrices.jl#L208

At first it looked to me like it had no axes checking, but then I realized that the check is buried inside the `convert` call there.  So if the vector is not something like `OneTo` then it throws a `DimensionMismatch`.

I think I will adopt the `convert` approach of `ToeplitzMatrices` and if y'all later figure out what would be a meaningful way to handle offsets then perhaps we can do that here too - if there is any use case!